presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushdownFilterIntoWindow.java

```diff
@@ -0,0 +1,184 @@
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.Session;
+import io.prestosql.matching.Capture;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.metadata.FunctionId;
+import io.prestosql.metadata.Metadata;
+import io.prestosql.spi.predicate.Domain;
+import io.prestosql.spi.predicate.Range;
+import io.prestosql.spi.predicate.TupleDomain;
+import io.prestosql.spi.predicate.ValueSet;
+import io.prestosql.sql.ExpressionUtils;
+import io.prestosql.sql.planner.DomainTranslator;
+import io.prestosql.sql.planner.PlanNodeIdAllocator;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.TypeProvider;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.FilterNode;
+import io.prestosql.sql.planner.plan.TopNRowNumberNode;
+import io.prestosql.sql.planner.plan.ValuesNode;
+import io.prestosql.sql.planner.plan.WindowNode;
+import io.prestosql.sql.tree.BooleanLiteral;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.QualifiedName;
+
+import java.util.Optional;
+import java.util.OptionalInt;
+
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.prestosql.SystemSessionProperties.isOptimizeTopNRowNumber;
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.spi.predicate.Marker.Bound.BELOW;
+import static io.prestosql.spi.predicate.Range.range;
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.sql.planner.DomainTranslator.fromPredicate;
+import static io.prestosql.sql.planner.plan.Patterns.filter;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static io.prestosql.sql.planner.plan.Patterns.window;
+import static java.lang.Math.toIntExact;
+
+public class PushdownFilterIntoWindow
+        implements Rule<FilterNode>
+{
+    private final Capture<WindowNode> childCapture;
+    private final Pattern<FilterNode> pattern;
+
+    private final Metadata metadata;
+    private final DomainTranslator domainTranslator;
+    private final FunctionId rowNumberFunctionId;
+
+    public PushdownFilterIntoWindow(Metadata metadata)
+    {
+        this.metadata = metadata;
+        this.domainTranslator = new DomainTranslator(metadata);
+        this.rowNumberFunctionId = metadata.resolveFunction(QualifiedName.of("row_number"), ImmutableList.of()).getFunctionId();
+        this.childCapture = newCapture();
+        this.pattern = filter()
+                .with(source().matching(window()
+                        .matching(window -> window.getWindowFunctions().size() == 1)
+                        .matching(window -> getOnlyElement(window.getWindowFunctions().values()).getResolvedFunction().getFunctionId().equals(rowNumberFunctionId))
+                        .capturedAs(childCapture)));
```

Comment:
I think there's one condition missing in the pattern: `.matching(window -> window.getOrderingScheme().isPresent())`

Reply:
I don't think so. `PushdownFilterIntoWindow` needs to return an empty values node even if `windowNode.getOrderingScheme().isPresent() == false`. It's handled in L.110.